### PR TITLE
Test TruffleRuby in CI

### PR DESCRIPTION
**Pull request by eregon**
# Description:

We'd like to test TruffleRuby in RubyGems' CI, but it's not working yet.
This is related to #2700, but with additional fixes to run tests.

cc @hsbt 

I will abide by the [code of conduct](https://github.com/rubygems/rubygems/blob/master/CODE_OF_CONDUCT.md).

### CODE REVIEW

**Comment by eregon** [comment]
Is there an easy way to disable AppVeyor from from building this branch?
It's not useful while I'm debugging, and probably makes other builds wait.

**Comment by eregon** [comment]
I rebased this and it should now use the latest 19.1.0 TruffleRuby release, but TravisCI is down, oh well.

**Comment by duckinator** [comment]
@eregon it looks like [your change to TruffleRuby last month](https://github.com/oracle/truffleruby/commit/e702cd31fbeebd59e75643b64a3cf8035112a76a) addresses a significant number of the errors in the last Travis CI run. It also appears that [19.1.1](https://github.com/oracle/truffleruby/releases/tag/vm-19.1.1) was released after that commit.

So — assuming your changes are in TruffleRuby 19.1.1 — I think if you rebase this, it may have fewer failures.

**Comment by eregon** [comment]
@duckinator 19.1.1 only contains security fixes and since there was none for TruffleRuby, it's exactly the same as the 19.1.0 TruffleRuby standalone. So, we have to wait the 19.2.0 release to get that fix. I don't think that fix alone will be enough.

Please let me work on this, since I know most about TruffleRuby and I think right now there is little the RubyGems can do to help.
@hsbt Until this is integrated, feel free to ignore TruffleRuby for RubyGems releases.

**Comment by eregon** [comment]
What we really need to make better progress on this is nightly builds (https://github.com/oracle/truffleruby/issues/1483). I'm trying to get those, but it takes a while to put all the pieces together.

It's rather frustrating that RubyGems tests mostly just work when we run a good part of them in TruffleRuby's CI, but for some reason it's a lot harder to make them run in TravisCI.

**Comment by duckinator** [comment]
@eregon understood. :+1:

**Comment by eregon** [comment]
Making some progress, the tests were somehow expecting `#{RUBY_ENGINE.upcase}_VERSION` (i.e., `TRUFFLERUBY_VERSION`) to be defined and hiding the error.
https://github.com/rubygems/rubygems/pull/2864 and https://github.com/bundler/bundler/pull/7275 solve that, please review those.

**Comment by hsbt** [comment]
@eregon Can you update with the current master branch? We moved to GitHub Actions instead of Travis CI. See https://github.com/rubygems/rubygems/blob/master/.github/workflows/ubuntu-rvm.yml#L16

**Comment by eregon** [comment]
Making progress here, by running it locally and fixing the missing `Errno::ENOTSUP` I get:
```
  1) Failure:
TestGemPackageTarReaderEntry#test_full_name_null [/home/eregon/code/rubygems/test/rubygems/test_gem_package_tar_reader_entry.rb:82]:
Gem::Package::TarInvalidError expected but nothing was raised.

  2) Failure:
TestGemRequire#test_activate_via_require_respects_loaded_files [/home/eregon/code/rubygems/test/rubygems/test_require.rb:242]:
benchmark should have already been loaded

  3) Failure:
TestGemRequire#test_realworld_default_gem [/home/eregon/code/rubygems/test/rubygems/test_require.rb:395]:
Expected "" to not be empty.

  4) Failure:
TestGemRequire#test_no_kernel_require_in_warn_with_uplevel [/home/eregon/code/rubygems/test/rubygems/test_require.rb:528]:
Expected /main\.rb:1: warning: uplevel\ntest\n$/ to match "main.rb:1:in `require': cannot load such file -- sub (LoadError)\n" +
"\tfrom main.rb:1:in `<main>'\n".

  5) Failure:
TestGemRequire#test_no_kernel_require_in_Kernel_warn_with_uplevel [/home/eregon/code/rubygems/test/rubygems/test_require.rb:528]:
Expected /main\.rb:1: warning: uplevel\ntest\n$/ to match "main.rb:1:in `require': cannot load such file -- sub (LoadError)\n" +
"\tfrom main.rb:1:in `<main>'\n".

  6) Failure:
TestStubSpecification#test_contains_requirable_file_eh_extension [/home/eregon/code/rubygems/test/rubygems/test_gem_stub_specification.rb:79]:
--- expected
+++ actual
@@ -1,2 +1 @@
-"Ignoring stub_e-2 because its extensions are not built. Try: gem pristine stub_e --version 2
-"
+""

  7) Skipped:
TestGemExtBuilder#test_build_extensions_extconf_bad [/home/eregon/code/rubygems/test/rubygems/test_gem_ext_builder.rb:244]:
Gem.ruby is not the name of the binary being run in the end

  8) Failure:
TestGem#test_self_try_activate_missing_extensions [/home/eregon/code/rubygems/test/rubygems/test_gem.rb:1305]:
--- expected
+++ actual
@@ -1,2 +1 @@
-"Ignoring ext-1 because its extensions are not built. Try: gem pristine ext --version 1
-"
+""

  9) Failure:
TestGemInstallUpdateOptions#test_add_install_update_options [/home/eregon/code/rubygems/test/rubygems/test_gem_install_update_options.rb:36]:
Expected false to be truthy.

 10) Failure:
TestGemPackage#test_verify_corrupt [/home/eregon/code/rubygems/test/rubygems/test_gem_package.rb:906]:
--- expected
+++ actual
@@ -1 +1 @@
-"tar is corrupt, name contains null byte in /tmp/corrupt20200307-16910-1xnrtgu"
+"package metadata is missing in /tmp/corrupt20200307-16910-1xnrtgu"

 11) Failure:
TestGemSpecification#test_contains_requirable_file_eh_extension [/home/eregon/code/rubygems/test/rubygems/test_gem_specification.rb:1695]:
--- expected
+++ actual
@@ -1,2 +1 @@
-"Ignoring ext-1 because its extensions are not built. Try: gem pristine ext --version 1
-"
+""

 12) Error:
TestGemSpecification#test_self_load_tainted:
SecurityError: Setting $SAFE is no longer supported.
    /home/eregon/code/rubygems/test/rubygems/test_gem_specification.rb:955:in `block in <top (required)>'
    /home/eregon/code/rubygems/test/rubygems/test_gem_specification.rb:955:in `block in test_self_load_tainted'

2155 runs, 6732 assertions, 10 failures, 1 errors, 1 skips
```

**Comment by eregon** [comment]
Latest run in CI:
```
  1) Skipped:
TestGemExtBuilder#test_build_extensions_extconf_bad [/home/runner/work/rubygems/rubygems/test/rubygems/test_gem_ext_builder.rb:244]:
Gem.ruby is not the name of the binary being run in the end

  2) Failure:
TestGemRequire#test_realworld_default_gem [/home/runner/work/rubygems/rubygems/test/rubygems/test_require.rb:395]:
Expected "" to not be empty.

  3) Failure:
TestGemRequire#test_no_kernel_require_in_Kernel_warn_with_uplevel [/home/runner/work/rubygems/rubygems/test/rubygems/test_require.rb:527]:
Expected /main\.rb:1: warning: uplevel\ntest\n$/ to match "main.rb:1:in `raise_load_error': cannot load such file -- sub (LoadError)\n" +
"\tfrom main.rb:1:in `require'\n" +
"\tfrom main.rb:1:in `<main>'\n".

  4) Failure:
TestGemRequire#test_activate_via_require_respects_loaded_files [/home/runner/work/rubygems/rubygems/test/rubygems/test_require.rb:242]:
benchmark should have already been loaded

  5) Failure:
TestGemRequire#test_no_kernel_require_in_warn_with_uplevel [/home/runner/work/rubygems/rubygems/test/rubygems/test_require.rb:527]:
Expected /main\.rb:1: warning: uplevel\ntest\n$/ to match "main.rb:1:in `raise_load_error': cannot load such file -- sub (LoadError)\n" +
"\tfrom main.rb:1:in `require'\n" +
"\tfrom main.rb:1:in `<main>'\n".

  6) Failure:
TestGemCommandsUpdateCommand#test_execute_system_specific_older_than_3_2_removes_plugins_dir [/home/runner/work/rubygems/rubygems/test/rubygems/test_gem_commands_update_command.rb:177]:
Plugins folder not removed when updating rubygems to pre-3.2.
Expected path '/home/runner/work/rubygems/rubygems/tmp/test_rubygems_3377/gemhome/plugins' to not exist.

  7) Failure:
TestGemCommandsUpdateCommand#test_execute_system_at_latest [/home/runner/work/rubygems/rubygems/test/rubygems/test_gem_commands_update_command.rb:99]:
Gem::MockGemUi::SystemExitException expected but nothing was raised.

  8) Failure:
TestGemCommandsUpdateCommand#test_execute_system_with_gems [/home/runner/work/rubygems/rubygems/test/rubygems/test_gem_commands_update_command.rb:230]:
Gem::MockGemUi::TermError expected but nothing was raised.

  9) Failure:
TestGemCommandsUpdateCommand#test_execute_system [/home/runner/work/rubygems/rubygems/test/rubygems/test_gem_commands_update_command.rb:82]:
Expected: "Updating rubygems-update"
  Actual: nil

 10) Failure:
TestGemCommandsUpdateCommand#test_execute_system_multiple [/home/runner/work/rubygems/rubygems/test/rubygems/test_gem_commands_update_command.rb:129]:
Expected: "Updating rubygems-update"
  Actual: nil

 11) Failure:
TestGemCommandsUpdateCommand#test_execute_system_specifically_to_latest_version [/home/runner/work/rubygems/rubygems/test/rubygems/test_gem_commands_update_command.rb:219]:
Expected: "Updating rubygems-update"
  Actual: nil

 12) Failure:
TestGemCommandsUpdateCommand#test_execute_system_specific [/home/runner/work/rubygems/rubygems/test/rubygems/test_gem_commands_update_command.rb:155]:
Expected: "Updating rubygems-update"
  Actual: nil

2170 runs, 6753 assertions, 11 failures, 0 errors, 1 skips
```
Looks like some new failures or new failing tests in TestGemCommandsUpdateCommand, and other than that just 4 failures left.

**Comment by deivid-rodriguez** [comment]
Thanks so much for your work on this @eregon :+1:.

**Comment by eregon** [comment]
Thanks, it's in large part due to @bjfish too :)

It's now down to 2 failures, `test_activate_via_require_respects_loaded_files` was quite hard to understand so I clarified by adding some comments and improve the logic in that test.
And `test_realworld_default_gem` is a known TruffleRuby bug I'll take a look at.

**Comment by eregon** [comment]
This is now passing, TruffleRuby now passes all RubyGems tests :tada: :tada: :tada: 

This is thanks to @bjfish and my work which fixed many things in TruffleRuby and also spent a fair amount of time understanding what some of the tests try to do and why they failed, which took a long time (>1 year since #2700) to get everything passing.

https://github.com/rubygems/rubygems/runs/658766102?check_suite_focus=true
```
2177 runs, 6807 assertions, 0 failures, 0 errors, 0 skips
```
And that's with __*zero*__ test excluded!

There are a few improvements to existing tests to be more reliable in this PR, as well as an added test for upgraded default gems as the existing test was not testing much.

Unfortunately these tests take a long time on TruffleRuby, I would think mostly because they use a lot of subprocesses and load quite a bit of code (+ of course tests are typically worst-case for a JIT, running a lot of different code just once). We're generally working on improving loading code speed in TruffleRuby so this should improve over time.
I don't think this should be much of an issue in practice though, this run takes ~30min, but each of the Bundler test suite runs takes ~15min, and there are many of them, and windows bundler tests take > 30min. So it shouldn't change much about total CI time and throughput.

@deivid-rodriguez @hsbt Could you review and merge?

File: .github/workflows/ubuntu.yml [hunk 1]
```diff
@@ -12,16 +12,18 @@ jobs:
     runs-on: ubuntu-latest
     strategy:
       matrix:
-        ruby: [ 2.3.8, 2.4.10, 2.5.8, 2.6.6, 2.7.1, jruby-9.2.11.1 ]
+        ruby: [ 2.3.8, 2.4.10, 2.5.8, 2.6.6, 2.7.1, jruby-9.2.11.1, truffleruby-head ]
```
Comment (hsbt) [inline .github/workflows/ubuntu.yml:15]: Should we test with the stable version of truffleruby?

Comment (eregon) [reply .github/workflows/ubuntu.yml:15]: The latest release doesn't pass the tests, so it's not an option.

I'll make sure to run RubyGems tests in TruffleRuby's CI too (already the case, but some tests are excluded and no longer need to be).
TruffleRuby's `master` (and therefore `truffleruby-head`) is only updated if the entire TruffleRuby CI passes, so it should not cause failures.

Comment (hsbt) [reply .github/workflows/ubuntu.yml:15]: I see. We can add the stable version of TruffleRuby after releasing from the current master version.

**Pull request merged**